### ISSUE TITLE
OsmAnd+ takes very long to refresh screen

### ISSUE THREAD

**Issue by speleoluc** [osmandapp/OsmAnd#18302]
### Description

Whenever I bring the OsmAnd+ screen back to top, even if I haven't changed position, I either get the splash screen or a white screen. After that, Android warns me the app is taking too long to answer and asks whether I want to wait or kill it. When It doesn't crash (...), it slowly brings up the map, tile by tile. By comparison, Gaia GPS comes up almost immediately.

### Steps to reproduce

Put OsmAnd+ in the background and then tap on its tile.

### Actual result

Long delay before the app crashes or displays its map.

### Expected result

Quick come back.

### Your Environment (required)

OsmAnd Version: 4.5.10
Android/iOS version: Android 11
Device model: Samsung Galaxy S5e tablet
Crash-Logs: ?
```

**Comment by DmitryAlexei** [comment]
4.6.0#842m 4600 
08.10.2023
- I opened OsmAnd
- then opened other apps, so OsmAnd was put in background
- after about 30-40 minutes of been in background, I put OsmAnd on top, and it crashed and proposed me to send exception log

The last recording in [exception.log](https://github.com/osmandapp/OsmAnd/files/12841610/exception.log) is from 02.10.23 and it contains the records: `java.lang.RuntimeException: An error occurred while executing doInBackground()`

**Comment by speleoluc** [comment]
The only exception.log file I find on my tablet is iin the osmand directory on the SD card:

Version  OsmAnd 1.9.4
21.12.2014 9:44:29
Apk Version : 1.9.4 196
Exception occured in thread Thread[main,5,main] : 
android.view.WindowManager$BadTokenException: Unable to add window -- token android.os.BinderProxy@42526030 is not valid; is your activity running?
	at android.view.ViewRootImpl.setView(ViewRootImpl.java:532)
	at android.view.WindowManagerGlobal.addView(WindowManagerGlobal.java:259)
	at android.view.WindowManagerImpl.addView(WindowManagerImpl.java:69)
	at android.app.Dialog.show(Dialog.java:286)
	at net.osmand.plus.ProgressImplementation$1.handleMessage(ProgressImplementation.java:54)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loop(Looper.java:136)
	at android.app.ActivityThread.main(ActivityThread.java:5001)
	at java.lang.reflect.Method.invokeNative(Native Method)
	at java.lang.reflect.Method.invoke(Method.java:515)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
	at dalvik.system.NativeStart.main(Native Method)

but it doesn't seem to date from 4.5.10.

I have other unsent emails with such logs, but Gmail won't send them and I can't download nor view the attached exception.log file.

**Comment by RenasAzad** [comment]
Device : zeus
Brand : Xiaomi
Model : 2201122G
Product : zeus_global
Build : TKQ1.220807.001 test-keys
Version : 13
App Version : OsmAnd+
Apk Version : 4.6.5 4605

hi all

the osmand+ takes a long time to be load (25 seconds)!

regards.

**Comment by aSemy** [comment]
Are there any workarounds for this issue while we wait for a fix? At the moment I'm finding OsmAnd to be practically unusable because it locks up and freezes every time I switch apps, like playing music or checking my messages or looking at a website. It's not fun to have to stop my journey for 5 minutes to repeatedly force stop OsmAnd and until it eventually opens properly.

**Comment by AndyB69** [comment]
@vshcherb Should I open a new issue for my variation of this thread? In my case:

- No crash is happening
- But, every time OsmAnd comes to the foreground, **it first blanks itself**, sometimes for up to 1 second, then redraws all the layers. That blanking on comign to foreground, with no zooming taking place, seems unneccessary - other mapping applications don't do that, and it's quite a usability issue, impacting the ability to make fast decisions using the map - the most common use case is when the screen is off, and you turn it on, the whole map is blanked and you have to wait for all the layers to redraw, but it hangs around on the blank map for 

The blank screen often lasts long enough to capture it - not milliseconds, but often half a second or more:
![Screenshot_20231210-163320](https://github.com/osmandapp/OsmAnd/assets/68859219/23c155a7-e62f-4615-9cbb-b4d3be824b43)

Here is the video again: [https://photos.app.goo.gl/dUsU8spgfB1s4qYf7]

**Comment by vshcherb** [comment]
Nightly: 15 (2-3) seconds to refresh / 25 (15) seconds to start
Release: 1-2 (< 1) seconds to refresh / 5-7 (3 seconds) to start

**Comment by AndyB69** [comment]
@vshcherb Thanks, though can you explain that update in layman's terms? Are you saying that the nightly release is currently those stats, but the next (?) release (4.6?) is showing a significant speed increase in map drawing?

**Comment by vshcherb** [comment]
Nope, we investigated that we have 2 separate issues:
1. Long app start itself - that needs to be improved in future
2. App refresh after moving from background (GPU memory refresh... ?) - we could investigate whether we could remove it somehow

**Comment by AndyB69** [comment]
Ah, thank you.

IMHO, Item 2 is the bigger issue - I've not seen other mapping apps do this on my phone (blank/redraw after moving from background) and it's a real usability hit in fairly common situations - imagine riding a bike, screen is off because you've been on a long straight road for a while, but you know a turn is coming, you see a side road coming up, late, turn the screen on check, but then have to wait 2 secs and have shot past the road by the time the map confirms it was the one to take - stuff like that :) Myself, if starting the app from nothing, I'm willing to wait a little longer. If other map apps don't blank/redraw when moving from background, does OsmAnd need to? It could be instantly available then, coming from background.

**Comment by vshcherb** [comment]
https://github.com/osmandapp/OsmAnd/pull/18764#issuecomment-1861709259 I think main issue (not  a "long cold start " issue) is fixed. So this one needs to be tested

**Comment by AndyB69** [comment]
Ah, see, this is one reason I love OsmAnd so much - in 4.6.9, the blanking/re-drawing whenever coming to the foreground has now been fixed. Beautiful. Thank you!! And all the best to the OsmAnd team for 2024!

**Comment by DmitryAlexei** [comment]
OsmAnd~ 4.7.0#1308m, released: 2024-01-03
No more blinking when evoking OsmAnd from the background. The map also does not redraw.

**Comment by JeckPeppinger** [comment]
F-Droid contains 4.5.10 version only ....

**Comment by yuriiurshuliak** [comment]
Please try to use the release version from Google Play.

**Comment by JeckPeppinger** [comment]
ah, i dont have google play. 

But I understand, you implement a bug and put out this version as the free one.
Just to lure people into the paid version.
Good idea! Great community, thanks for the info!

**Comment by speleoluc** [comment]
Developers don't implement bugs. They're accidents. But if the fix is in
the free version, what lures you into buying the paid version?


-- 
Luc Le Blanc

Le mer. 24 janv. 2024 07 h 38, JeckPeppinger ***@***.***> a
écrit :

**Comment by speleoluc** [comment]
As a software developer (not on Osmand) with a C++ background, I'd have
thought the code base is the same for all versions, if only for practical
reasons, the paid options merely being filtered by compilation options.


-- 
Luc Le Blanc

Le mer. 24 janv. 2024 08 h 11, JeckPeppinger ***@***.***> a
écrit :

**Comment by JeckPeppinger** [comment]
ah, ok, i cant say about the code base is same for both distribution channels (google play, F-droid). 
However, the available version on F-Droid is ages old, and very strangely introduced a massive and obvious bug.
Uploading an apk to F-Droid should not be a big deal, so that the bug is out is quite obvious on purpose.
Purpose being to end the F-Droid distribution? And lure users into the paid one .... ?

**Comment by JeckPeppinger** [comment]
not sure what a recipe is, but seems to be 50 lines of configuration where most of them are completely unchanged to the previous version.

However, its clear that the search for a new tool is on as it appears osmand is now commercialising. Well, fair enough, they can do that, good tool. First create a large user base, use free maps and other free resources, many developers and contributers who work for free and then just go the cash path. Nice move!

**Comment by vshcherb** [comment]
Just tested with 4.7 and map constantly refreshes if you switch between app even if you switch between the same app. That what was discussed if I remeber to keep the same
